Sources/OpenCombine/Publishers/Publishers.Zip.swift

```diff
@@ -0,0 +1,756 @@
+                if case .stopProcessing = work {
+                    queueIsBeingProcessed = false
+                    processingValueForChild = nil
+                    receiveValueDemandOverride = demandReceivedWhileProcessing
+                    demandReceivedWhileProcessing = nil
+                }
+                return work
+            }
+
+            switch action {
+            case .stopProcessing:
+                return receiveValueDemandOverride
+            case .noAction:
+                break
+            case .sendFinishDownstream:
+                downstream.receive(completion: .finished)
+                let activeChildren = lock.do {
+                    upstreamSubscriptions.filter { $0.state == .active }
+                }
+                activeChildren.forEach { $0.cancel() }
+                return receiveValueDemandOverride
+            case .sendValueDownstream(let value):
+                let newDemand = downstream.receive(value)
+                if newDemand != .none {
+                    lock.do {
+                        downstreamDemand += newDemand
+                        demandReceivedWhileProcessing = newDemand
+                    }
+                }
+            case .sendRequestUpstream(let demand):
+                lock.do { upstreamSubscriptions
+                    .filter { $0.childIndex != processingValueForChild?.childIndex }
+                }.forEach { $0.request(demand) }
+            }
+        }
+    }
+}
+
+extension InnerBase: Subscription {
+    fileprivate final func request(_ demand: Subscribers.Demand) {
+        guard demand != .none else {
+            fatalError()
+        }
+        let shouldProcessQueue: Bool = lock.do {
+            downstreamDemand += demand  // TODO: Move this to the action processing?
+            queuedWork.append(.receivedRequestFromDownstream(demand: demand))
+            if queueIsBeingProcessed {
+                demandReceivedWhileProcessing = demand
+            } else {
+                queueIsBeingProcessed = true
+                return true
+            }
+            return false
+        }
+
+        if shouldProcessQueue {
+            processQueue()
+        }
+    }
+
+    fileprivate final func cancel() {
+        lock.do { upstreamSubscriptions }.forEach { $0.cancel() }
+    }
+}
+
+private enum ChildState {
+    case waitingForSubscription
+    case active
+    case finished
+    case failed
+    case canceled
+}
+
+// Note that it's critical that this protocol not have any associated types - specifically
+// note that it does not refer to `Upstream`.
+// This allows `InnerBase` to do most of the heavy lifting without regard to the
+// upstream publisher's value type.
+private protocol ChildSubscription: AnyObject, Subscription {
+    var state: ChildState { get set }
+    var childIndex: Int { get }
+    var hasValue: Bool { get }
+}
+
+fileprivate final class ChildSubscriber<Upstream: Publisher, Downstream: Subscriber>
```

Comment:
I had a look at `SwitchToLatest` and I'm a bit confused, shouldn't the `Outer` class there be named `Inner` or maybe some other name? The current name leads to this, not the mismatch between argument name and argument type:

```swift
init(inner: Outer, index: UInt64) {
  self.index = index
  self.outer = inner
}
```

I think these names (applicable to `Side` too) don't make their purpose incredibly clear. What exactly is being `Outer` (or `Inner`, or `Side` for that matter) here? A subscription? A subscriber? A publisher?

Reply:
(I don't mind converting that class to a struct, but it's not as straightforward and I'd like to know your opinion on these issues first so that I can implement it in a better way)

Reply:
I'm still not sure what would be the purpose of changing that class to a struct and the naming still seems to be inconsistent to me. Can you please elaborate how you'd like it to look and what would be the best naming approach consistent across the codebase?

Reply:
> I'm still not sure what would be the purpose of changing that class to a struct and the naming still seems to be inconsistent to me. Can you please elaborate how you'd like it to look and what would be the best naming approach consistent across the codebase?

Maybe the best practice is to align with Combine's naming.